Desktop e-book: Chapter 1 & 2 (#16149)
dotnet/docs: docs/architecture/modernize-desktop/whats-new-with-netcore3- for-desktop.md

> # What&#39;s new with .NET Core 3.0 for Desktop?

Reviewer: - Why there is no metadata header for this file ?
- This file name contains a space, it should be removed.
- Why encode thr `'` character ? I don't think there is a problem putting it directly.

   ```suggestion
   # What's new with .NET Core 3.0 for Desktop?
   ```

> 

Reviewer 2: file name here is problematic too. I'd avoid long names and specific versions listed.

Suggestion: whats-new-dotnet-core-desktop.md

> ## From .NET Framework to .NET Core. The motivation behind .NET **Core**

Reviewer 2: ```suggestion
## From .NET Framework to .NET Core - The motivation behind .NET Core
```

> Since its inception in 2002, .NET Framework has evolved through the years to support many technologies like Windows Desktop, ASP.NET, Entity Framework, Windows Store and many others. All of them are very different in nature. Therefore, Microsoft was approaching this evolution by taking parts of the .NET Framework and creating a different application stack for each technology. That way, development capabilities could be customized for the needs of the specific stack, which maximized the potential of every platform. Of course, that lead to fragmentation on the versions of the .NET Framework maintained by different independent teams. All these stacks have a common structure, containing an App Model, a Framework and a Runtime, but they differ in the implementation of each of these parts.

Reviewer: I think `Windows Forms` is the correct name.

```suggestion
Since its inception in 2002, .NET Framework has evolved through the years to support many technologies like Windows Forms, ASP.NET, Entity Framework, Windows Store and many others. All of them are very different in nature. Therefore, Microsoft was approaching this evolution by taking parts of the .NET Framework and creating a different application stack for each technology. That way, development capabilities could be customized for the needs of the specific stack, which maximized the potential of every platform. Of course, that lead to fragmentation on the versions of the .NET Framework maintained by different independent teams. All these stacks have a common structure, containing an App Model, a Framework and a Runtime, but they differ in the implementation of each of these parts.
```

> ![](./media/2-1.png)

Reviewer: linter requires alt text.

By the time Windows 8 was released, the concept of the Portable Class Libraries (PCL) was born. Originally, the .NET Framework was designed around the assumption that it is always deployed as a single unit, so [factoring](http://en.wikipedia.org/wiki/Decomposition_(computer_science)) was not a concern. To face the problem of code sharing between verticals, the driving force was on how to refactor the framework. The idea of contracts is to provide a well factored API surface area. Contracts are simply assemblies that you compile against and are design with proper factoring in mind taking care of the dependencies between them.

This leads to a reasoning about the API differences between verticals at the assembly level, as opposed to the individual API level that we had before. This aspect enabled a class library experience that can target multiple verticals, also known as portable class libraries.

> ![](./media/2-2.png)

Reviewer: alt text ?

> With PCL the experience of development is unified across verticals base on the API shape and the most pressing need to create libraries running on different verticals is addressed. But there is still a great challenge: APIs are only portable when the implementation is moved forward across all the verticals, and still verticals have independent implementations.

Reviewer: ```suggestion
With PCL, the experience of development is unified across verticals base on the API shape and the most pressing need to create libraries running on different verticals is addressed. But there is still a great challenge: APIs are only portable when the implementation is moved forward across all the verticals, and still verticals have independent implementations.
```

> A better approach is to unify the implementations across verticals by providing a well factored implementation instead of a well factored view. It is a lot simpler to ask each team owning a specific component to think about how their APIs work across all verticals than trying to retroactively provide a consistent API stack on top. This is where .NET Standard comes in, see details on next section.

Reviewer: ```suggestion
A better approach is to unify the implementations across verticals by providing a well-factored implementation instead of a well-factored view. It is a lot simpler to ask each team owning a specific component to think about how their APIs work across all verticals than trying to retroactively provide a consistent API stack on top. This is where .NET Standard comes in, see details on next section.
```

To start with, it is difficult for application developers to take a dependency on a recently released framework. You have either to take a dependency on the latest OS or provide an application installer that will install the .NET Framework along with the application. If you are a web developer, you might not even have this option as the IT department establishes the server supported version.

Even if you are willing to go through the trouble of providing an installer in order to chain in the .NET Framework setup, you may find that upgrading the .NET Framework can break other applications.

Despite the efforts to provide backward compatible versions of the framework, there are compatible changes that can break applications. For example, adding an interface to an existing type that can change how this type is serialized and cause breaking problems depending on the existing code. Because NET Framework installed base is huge, fighting against these breaking scenarios slows down the pace of innovations inside the .NET Framework.

To solve all these issues Microsoft has developed .NET Core to approach the evolution of the .NET Platform.

## **Introduction to .NET Core**

> The .NET Core is the evolution of Microsoft&#39;s .NET technology into a modular, cross platform, open source and cloud ready platform, which runs on Windows, Mac, and Linux with plans to run also on ARM based architectures like Android and IoT.

Reviewer: ```suggestion
The .NET Core is the evolution of Microsoft's .NET technology into a modular, cross platform, open source, and cloud ready platform, which runs on Windows, macOS, and Linux with plans to run also on ARM based architectures like Android and IoT.
```

> The purpose of .NET Core is to provide a unified platform for all types of applications, which includes Windows, cross platform and mobile applications. [.NET Standard library](https://docs.microsoft.com/en-us/dotnet/articles/standard/library) enables this by providing shared base API&#39;s, which every application model needs, and excluding any application model specific API&#39;s.

Reviewer: ```suggestion
The purpose of .NET Core is to provide a unified platform for all types of applications, which includes Windows, cross platform and mobile applications. [.NET Standard library](../../standard/net-standard.md) enables this by providing shared base APIs, which every application model needs, and excluding any application model-specific API.
```

This framework gives applications many benefits in terms of efficiency and performance, simplifying the packaging and deployment in the different supported platforms.

The benefits of .NET Core come from these three characteristics:

- **Cross-platform:** It allows application execution on different platforms, Windows, macOS and Linux.

> - **Open source** : .NET Core platform is open source and available through GitHub, fostering transparency and community contributions.

Reviewer: To be consistent with the previous item.

```suggestion
- **Open source:** .NET Core platform is open source and available through GitHub, fostering transparency and community contributions.
```

> - **Strongly supported** : Microsoft officially supports .NET Core.

Reviewer: ```suggestion
- **Strongly supported:** Microsoft officially supports .NET Core.
```

In .NET Core 3.0, there is planned support for developing the following application types:

> ![](./media/2-3.png)

Reviewer: alt text ?

> The goal for this framework is quite impressive: to target every type of .NET development present and future including Desktop, Web, Cloud, Mobile, Gaming, IoT and AI. Microsoft plans to complete this vision with .NET 5 at the end of 2020. Note that the &quot;Core&quot; name was removed to reinforce its uniqueness in the .NET World.

Reviewer: ```suggestion
The goal for this framework is quite impressive: to target every type of .NET development present and future including Desktop, Web, Cloud, Mobile, Gaming, IoT and AI. Microsoft plans to complete this vision with .NET 5 at the end of 2020. Note that the "Core" name was removed to reinforce its uniqueness in the .NET World.
```

> ![](./media/2-4.png)

Reviewer: alt text ?

## **.NET Framework vs .NET Core**

Therefore, you now understand the relevance of .NET Core inside the Microsoft strategy for .NET and might be wondering what happens with .NET Framework and asking questions like: do you have to abandon it? Is it going to disappear? What are my choices to modernize the applications I have on .NET Framework?

> The good news is Microsoft plans to support both .NET Framework and .NET Core. In fact, the announcement of .NET Core 3.0 was followed by exciting features in .NET Framework regarding modernization, so let&#39;s take a brief look at both paths.

Reviewer: ```suggestion
The good news is Microsoft plans to support both .NET Framework and .NET Core. In fact, the announcement of .NET Core 3.0 was followed by exciting features in .NET Framework regarding modernization, so let's take a brief look at both paths.
```

> Listening to the .NET Framework developer&#39;s community suggestions Microsoft have addressed three main scenarios for **.NET Core** :

Reviewer: ```suggestion
Listening to the .NET Framework developer's community suggestions, Microsoft have addressed three main scenarios for **.NET Core**:
```

- **Side-by-side versions of .NET supporting Windows Forms and WPF** : This solves the problem of side effects when updating the machine&#39;s framework version. With .NET Core multiple versions can be installed on the same machine and for each application you can specify, which version of .NET Core you&#39;d like it to use. Even more, now you can develop and run Windows Forms and WPF on top of .NET Core.
- **Embed .NET directly into an application** : You can now ship the .NET Core as part of your application package. This enables you to take advantage of the latest version, features, and APIs without having to wait for a specific version to be installed on the machine.
- **Take advantage of .NET Core features** : .NET Core is the fast-moving, open source version of .NET. Its side-by-side nature enables fast introduction of new innovative APIs and BCL (Base Class Libraries) improvements without the risk of breaking compatibility. Now Windows Forms and WPF applications on Windows can take advantage of the latest .NET Core features, which also includes more fundamental fixes for runtime performance, high-DPI support, and so on.

> Regarding **.NET Framework 4.8, Microsoft** have addressedthree scenarios:

Reviewer: ```suggestion
Regarding **.NET Framework 4.8**, Microsoft have addressed three scenarios:
```

- **High DPI improvements** : The resolution of displays is increasing to 4K and 8K resolutions. We want to make sure your existing Windows Forms and WPF applications can look great on these new displays. That&#39;s why we are adding improvements for HDPI

Since .NET Framework is working on millions of machines, Microsoft will continue to support it but will not be adding new features.

.NET Core is the open source, cross-platform, and fast-moving version of .NET. Because of its side-by-side nature, it can take changes without the fear of breaking any application. This means that .NET Core will get new APIs and language features over time that .NET Framework cannot.

An essential part of this roadmap is to ease developers to move applications to .NET Core. .NET Core 3.0 takes a huge step by adding WPF, Windows Forms and Entity Framework 6 support, and we will keep porting APIs and features to help close the gap and make migration easier for those who chose to do so.

> So, if you have existing .NET Framework applications, you should not feel pressured to move to .NET Core. .NET Framework will be fully supported and will always be a part of Windows. However, if in the future you want to use the newest language features and APIs, you&#39;ll need to move your applications to .NET Core.

Reviewer: ```suggestion
So, if you have existing .NET Framework applications, you should not feel pressured to move to .NET Core. .NET Framework will be fully supported and will always be a part of Windows. However, if in the future you want to use the newest language features and APIs, you'll need to move your applications to .NET Core.
```

> For your brand-new desktop applications, we recommend starting directly on .NET Core. It is lightweight, runs side by ride, high performance and cross platform and fits perfectly on containers and microservices architectures.

Reviewer: ```suggestion
For your brand-new desktop applications, we recommend starting directly on .NET Core. It is lightweight, runs side-by-side, high performance and cross platform and fits perfectly on containers and microservices architectures.
```

> ![](./media/2-5.png)

Reviewer: alt text ?

The .NET Standard enables the following key scenarios:

- Defines uniform set of base class libraries APIs for all .NET implementations to implement, independent of the workload.
- Enables developers to produce portable libraries that are usable across .NET implementations, using this same set of APIs.
- Reduces or even eliminates conditional compilation of shared source due to .NET APIs, only for OS APIs.

.NET Standard is the evolution of PCLs and here are the main differences between .NET Standard and PCLs:

- .NET Standard is a set of curated APIs, picked by Microsoft, PCLs are not.
- The APIs that a PCL contains is dependent on the platforms that you choose to target when you create a PCL. This makes a PCL only sharable for the specific targets that you choose.

> - .NET Standard is platform-agnostic, it can run anywhere, on Windows, Mac, Linux and so on.

Reviewer: ```suggestion
- .NET Standard is platform-agnostic, it can run anywhere, on Windows, macOS, Linux and so on.
```

A new version of .NET Standard, version 2.1, was released at the same time. As expected, all new .NET Standard APIs are part of .NET Core 3.0.

You can build desktop applications with C#, F#, and VB with .NET Core 3.0.

Also, important to notice that both Windows Forms and WPF implementations for .NET Core 3 are being open sourced.

### XAML Islands

XAML Islands is a set of components for developers to use the new Windows 10 controls (UWP XAML controls) in their current WPF, Windows Forms, and native Win32 apps (like MFC). You can have your &quot;islands&quot; of UWP XAML controls wherever you want inside your Win32 apps.

These XAML Islands are possible because Windows 10 1903 update introduces a set of APIs that allows hosting UWP XAML content in Win32 windows using windows handlers (HWnds). Notice that only apps running on Windows 10 1903 and above can use XAML Islands.

To facilitate to create XAML Islands for Windows Forms and WPF developers, the Windows Community Toolkit introduces a set of .NET wrappers in several NuGet packages. Those wrappers are the Wrapped Controls and Hosting Controls:

> - The WebView, WebViewCompatible, [InkCanvas](https://docs.microsoft.com/windows/communitytoolkit/controls/wpf-winforms/inkcanvas), [MediaPlayerElement](https://docs.microsoft.com/windows/communitytoolkit/controls/wpf-winforms/mediaplayerelement), and [MapControl](https://docs.microsoft.com/en-us/windows/communitytoolkit/controls/wpf-winforms/mapcontrol) wrapped controls wrap some UWP XAML controls into Windows Forms or WPF controls, hiding UWP concepts for those developers.

Reviewer 2: ```suggestion
- The WebView, WebViewCompatible, [InkCanvas](/windows/communitytoolkit/controls/wpf-winforms/inkcanvas), [MediaPlayerElement](/windows/communitytoolkit/controls/wpf-winforms/mediaplayerelement), and [MapControl](/windows/communitytoolkit/controls/wpf-winforms/mapcontrol) wrapped controls wrap some UWP XAML controls into Windows Forms or WPF controls, hiding UWP concepts for those developers.
```
